### ISSUE TITLE
Save session for later use

### ISSUE THREAD

**Issue by mowaiskalam** [Festify/cordova-spotify#14]
I am using this plugin and it's working great. But i am stuck in this feature that i want to save session for later use. 

Actually my work is that in my app customer will come and connect their spotify. And then they can play song from their spotify account. So i am able to connect their spotify account and play. But i need to save that success session for later use. I know that this functionality is not implemented currently and will be available soon but i want this feature now.

Can you please help me to implement this feature. Please just inform me what i have to do in your plugin to implement this functionality. I know it's not an issue but i need it urgent i will be very thankful to you.

**Comment by NeoLegends** [comment]
Great! We pretty much have the same desire for this feature.

You're probably going to need to do something like this:

- Fork this
- Save the refresh token after the session has been created to local storage
- Add new methods in the native parts `src/ios` / `src/android` that accept a refresh token and use that to initialize the native Spotify SDK sessions
- Add a method that attempts to login with the saved refresh token (from local storage), and returns e.g. false or throws an exception, if no refresh token is currently saved
- Create a pull request to merge your changes back into the tree

We are probably going to do another Festify development session by tomorrow night, so maybe we will have the time to implement this ourselves by then.

**Comment by mowaiskalam** [comment]
@NeoLegends Thanks for the guild lines.

But the problem is i don't know the development of native sdk :( but i will do it in any way. 

Here i am confuse with the 3rd part you wrote that i have to add a method in the native parts that accept a refresh token.
The refresh token swap implementation is on our server. Should i have to re-initialize auth token from refresh if it's expired ? because both swapping function are on our server.  So i make these process on AngularJS side (cordova app) for swapping refresh tokens.

**Comment by NeoLegends** [comment]
Exactly, you will need to get a new authentication token via the refresh token. Whichever way works - in the end the Spotify SDKs have methods to do the reauth, we just need to provide them with the data they need. I do not know exactly which steps have to be taken in order to make this work - but the general direction is always the following: try to give the SDKs the data they need and try to minimize the implications for the user.

Well, you will have to learn native coding then.

**Comment by NeoLegends** [comment]
@mowaiskalam Android reauthing is now working, iOS will soon too. Please check out the 'feature/login'-branch.

**Comment by mowaiskalam** [comment]
Oh! Thank you @NeoLegends. (y)

**Comment by NeoLegends** [comment]
iOS functionality has been implemented as well.

**Comment by mowaiskalam** [comment]
Great! Should i have to use master branch now or 'feature/login' ?

**Comment by NeoLegends** [comment]
`master` is the branch the releases are drafted from, so it kinda is the "stable" version and you should use that. `develop` is the current state of development, so basically the "nightly" version.

`feature/login` was a temporary branch specifically for implementing the reauthing feature. It has been deleted now.

**Comment by mowaiskalam** [comment]
@NeoLegends I am implementing reauthing feature. Here i am confuse with the login function as it needs me to provide 'param' object again which have 'urlscheme, tokenRefreshUrl etc'.

Now the problem here i am facing is if i open the app after closing i have to call authenticate again then i can call login function anywhere in the app. Am i right ?

I am assuming that i have to provide a 'refresh_token' (which i will store in the database against user) and it will check that the session time expire (if yes then it will use 'refresh_token' to refresh session again). Let me know if i am going wrong ? Thanks

**Comment by NeoLegends** [comment]
Do not store the refresh token on your server! The plugin will automatically persist the refresh token during `authenticate`.

If you would like to sign a previous user back in, you call `login` with the `tokenRefreshUrl` and the `clientId`. The plugin will then attempt to do the reauth. If it succeeds, the promise will resolve to a session object, just like `authenticate`. Otherwise the promise will resolve to `null`, in which case you'll have to call `authenticate` to do the full login process.

Basically like this: 

```js
cordova.plugins.spotify.login({ tokenRefreshUrl, clientId })
    .then(session => session ? session : cordova.plugins.authenticate(<full params>))
    .then(session => console.log("Hey, I've got the session in any case now:", session));
```

or like this: https://github.com/Festify/app/blob/develop/elements/providers/playback/playback-spotify.html#L27

**Comment by mowaiskalam** [comment]
@NeoLegends Oh! That's great it's storing refresh token itself. 

I understand now. When i was trying this i have called 'authenticate' and then immediately 'login' for checking it throws me error of 'SPTAudioStreamingController' already in use when i debug and set error callback too in cordova-spotify.js, Is it because i should wait for the expiration of the token ?

**Comment by NeoLegends** [comment]
Yeah nah. That's just a bug that is waiting to be fixed. Right now only a single login can happen per "plugin run".

Login needs to be called first.

**Comment by mowaiskalam** [comment]
@NeoLegends 

Not sure if this below situation relates to the same bug above or if it is a different issue but this is what's happening.

If I authenticate a user , and log him in on subsequent app start up, its working fine. However if I have to remove the spotify account and have to add a different account to the app, its giving the same error of audio streaming controller in use.

I've tried using your logout function and then using authenticate again but didn't help.

**Comment by NeoLegends** [comment]
Yeah logging out isn't really supported at the moment either. The logout method is just a stub. Will come soon.

**Comment by mowaiskalam** [comment]
Can you guide me how to proceed here if I want to. I am kind of stuck in this app and I have to allow user to delete Spotify integration and connect again.

Do I need to make changes to the native code? or is it just the javascript part?

**Comment by NeoLegends** [comment]
You are going to need native code for this, since the Spotify player (which is native) needs to be signed out.

Unfortunately I will not guide you this time because this library is advertised as beta (so not fully released yet) and betas simply have rough edges here. Also, we (the Festify team) picked up Objective C / Java just by building this library, so I suggest you do so, too. The code is - albeit mostly undocumented - rather simple and should be pretty easy to understand. Fork the repo, read the Spotify SDK documentation, learn ObjC / Java, integrate the logout functionality, and submit a PR. This is how we did everything as well - learning by doing.

That said: we are going to build logout functionality eventually (since we're going to need it as well), we just don't know exactly when we have the time to do so.
